File: Aliucord/src/main/java/com/aliucord/entities/NotificationData.java
/*
 * Copyright (c) 2021 Juby210
 * Licensed under the Open Software License version 3.0
 */

package com.aliucord.entities;

import android.graphics.drawable.Drawable;
import android.view.View;

import androidx.annotation.IdRes;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.discord.app.AppComponent;
import com.discord.models.sticker.dto.ModelSticker;

import java.util.List;

import c0.e0.c;
import kotlin.Unit;
import kotlin.jvm.functions.Function1;

/**
 * Notification builder.
 */
@SuppressWarnings("unused")
public class NotificationData {
    private CharSequence title;
    private CharSequence subtitle;
    private CharSequence body;
    private Drawable attachmentBackground;
    private String attachmentUrl;
    private Drawable attachment;
    private List<ModelSticker> stickers;
    private String iconUrl = "";
    private Integer iconResId;
    private Drawable iconTopRight;
    private Integer autoDismissPeriodSecs = 5;
    private List<? extends c<? extends AppComponent>> validScreens;
    private Function1<? super View, Unit> onClickTopRightIcon;
    private Function1<? super View, Unit> onClick;

    /**
     * @return Title of the notification.
     * @see NotificationData#setTitle(CharSequence)
     */
    public CharSequence getTitle() {
        return title;
    }

    /**
     * Sets the title.
     * @param title Title.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setTitle(@Nullable CharSequence title) {
        this.title = title;
        return this;
    }

    /**
     * @return Subtitle of the notification.
     * @see NotificationData#setSubtitle(CharSequence)
     */
    public CharSequence getSubtitle() {
        return subtitle;
    }

    /**
     * Sets the subtitle.
     * @param subtitle Subtitle.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setSubtitle(@Nullable CharSequence subtitle) {
        this.subtitle = subtitle;
        return this;
    }

    /**
     * @return Body of the notification.
     * @see NotificationData#setBody(CharSequence)
     */
    public CharSequence getBody() {
        return body;
    }

    /**
     * Sets the body.
     * @param body Body.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setBody(@Nullable CharSequence body) {
        this.body = body;
        return this;
    }

    /**
     * @return Attachment background {@link Drawable} of the notification.
     * @see NotificationData#setAttachmentBackground(Drawable)
     */
    public Drawable getAttachmentBackground() {
        return attachmentBackground;
    }

    /**
     * Sets the background for the attachment.
     * @param attachmentBackground Background {@link Drawable} of the attachment.
     * @return {@link NotificationData} for chaining.
     * @see NotificationData#setAttachment(Drawable)
     * @see NotificationData#setAttachmentUrl(String)
     */
    public NotificationData setAttachmentBackground(@Nullable Drawable attachmentBackground) {
        this.attachmentBackground = attachmentBackground;
        return this;
    }

    /**
     * @return Attachment URL of the notification.
     * @see NotificationData#setAttachmentUrl(String)
     */
    public String getAttachmentUrl() {
        return attachmentUrl;
    }

    /**
     * Sets the attachment URL.
     * @param attachmentUrl URL of the attachment.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setAttachmentUrl(@Nullable String attachmentUrl) {
        this.attachmentUrl = attachmentUrl;
        return this;
    }

    /**
     * @return Attachment {@link Drawable} of the notification.
     * @see NotificationData#setAttachment(Drawable) (CharSequence)
     */
    public Drawable getAttachment() {
        return attachment;
    }

    /**
     * Sets the attachment.
     * @param attachment Attachment {@link Drawable}.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setAttachment(@Nullable Drawable attachment) {
        this.attachment = attachment;
        return this;
    }

    /**
     * @return Stickers of the notification.
     * @see NotificationData#setStickers(List)
     */
    public List<ModelSticker> getStickers() {
        return stickers;
    }

    /**
     * Sets the stickers.
     * @param stickers {@link List} of stickers.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setStickers(@Nullable List<ModelSticker> stickers) {
        this.stickers = stickers;
        return this;
    }

    /**
     * @return Icon URL of the notification.
     * @see NotificationData#setIconUrl(String)
     */
    public String getIconUrl() {
        return iconUrl;
    }

    /**
     * Sets the icon URL.
     * @param iconUrl URL of the icon.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setIconUrl(@NonNull String iconUrl) {
        this.iconUrl = iconUrl;
        return this;
    }

    /**
     * @return Title of the notification
     * @see NotificationData#setTitle(CharSequence)
     */
    public Integer getIconResId() {
        return iconResId;
    }

    /**
     * Sets the icon.
     * @param iconResId {@link IdRes} of the icon {@link Drawable}.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setIconResId(@Nullable Integer iconResId) {
        this.iconResId = iconResId;
        return this;
    }

    /**
     * @return {@link Drawable} of the icon in the top right corner of the notification.
     * @see NotificationData#setIconTopRight(Drawable)
     */
    public Drawable getIconTopRight() {
        return iconTopRight;
    }

    /**
     * Sets the icon on the top right corner of the notification.
     * @param iconTopRight {@link Drawable} of the icon.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setIconTopRight(@Nullable Drawable iconTopRight) {
        this.iconTopRight = iconTopRight;
        return this;
    }

    /**
     * @return Auto dismiss period time in seconds
     * @see NotificationData#setAutoDismissPeriodSecs(Integer)
     */
    public Integer getAutoDismissPeriodSecs() {
        return autoDismissPeriodSecs;
    }

    /**
     * Sets the auto dismiss period.
     * @param autoDismissPeriodSecs Dismiss period in seconds.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setAutoDismissPeriodSecs(@Nullable Integer autoDismissPeriodSecs) {
        this.autoDismissPeriodSecs = autoDismissPeriodSecs;
        return this;
    }

    //FIXME Documentation is invalid.
    /**
     * @return Valid screens
     * @see NotificationData#setValidScreens(List)
     */
    public List<? extends c<? extends AppComponent>> getValidScreens() {
        return validScreens;
    }

    //FIXME Documentation is invalid.
    /**
     * Sets valid screens for the notification.
     * @param validScreens Valid screens.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setValidScreens(@Nullable List<? extends c<? extends AppComponent>> validScreens) {
        this.validScreens = validScreens;
        return this;
    }

    /**
     * @return Block corresponding the onClick action of the icon in the top right corner of the notification.
     * @see NotificationData#setOnClickTopRightIcon(Function1)
     */
    public Function1<? super View, Unit> getOnClickTopRightIcon() {
        return onClickTopRightIcon;
    }

    /**
     * Sets the callback for the notification top right corner icon onClick action.
     * @param onClickTopRightIcon Block to execute after clicking the icon in the top right corner of the notification.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setOnClickTopRightIcon(@Nullable Function1<? super View, Unit> onClickTopRightIcon) {
        this.onClickTopRightIcon = onClickTopRightIcon;
        return this;
    }

    /**
     * @return Block corresponding the onClick action of the notification.
     * @see NotificationData#setTitle(CharSequence)
     */
    public Function1<? super View, Unit> getOnClick() {
        return onClick;
    }

    /**
     * Sets the callback for the notification onClick action.
     * @param onClick Block to execute after clicking the notification.
     * @return {@link NotificationData} for chaining.
     */
    public NotificationData setOnClick(@Nullable Function1<? super View, Unit> onClick) {
        this.onClick = onClick;
        return this;
    }
}

File: Aliucord/src/main/java/com/aliucord/entities/NotificationData.kt
/*
 * Copyright (c) 2021 Juby210 & Vendicated
 * Licensed under the Open Software License version 3.0
 */
package com.aliucord.entities

import android.graphics.drawable.Drawable
import android.view.View
import com.discord.api.sticker.Sticker
import com.discord.app.AppComponent
import d0.e0.c

/**
 * Notification builder.
 */
@Suppress("unused", "MemberVisibilityCanBePrivate")
class NotificationData {
    /**
     * @return Title of the notification.
     * @see NotificationData.setTitle
     */
    var title: CharSequence? = null
        private set

    /**
     * @return Subtitle of the notification.
     * @see NotificationData.setSubtitle
     */
    var subtitle: CharSequence? = null
        private set

    /**
     * @return Body of the notification.
     * @see NotificationData.setBody
     */
    var body: CharSequence? = null
        private set

    /**
     * @return Attachment background [Drawable] of the notification.
     * @see NotificationData.setAttachmentBackground
     */
    var attachmentBackground: Drawable? = null
        private set

    /**
     * @return Attachment URL of the notification.
     * @see NotificationData.setAttachmentUrl
     */
    var attachmentUrl: String? = null
        private set

    /**
     * @return Attachment [Drawable] of the notification.
     * @see NotificationData.setAttachment
     */
    var attachment: Drawable? = null
        private set

    /**
     * @return Stickers of the notification.
     * @see NotificationData.setStickers
     */
    var stickers: List<Sticker>? = null
        private set

    /**
     * @return Icon URL of the notification.
     * @see NotificationData.setIconUrl
     */
    var iconUrl = ""
        private set

    /**
     * @return Title of the notification
     * @see NotificationData.setTitle
     */
    var iconResId: Int? = null
        private set

    /**
     * @return [Drawable] of the icon in the top right corner of the notification.
     * @see NotificationData.setIconTopRight
     */
    var iconTopRight: Drawable? = null
        private set

    /**
     * @return Auto dismiss period time in seconds
     * @see NotificationData.setAutoDismissPeriodSecs
     */
    var autoDismissPeriodSecs: Int? = 5
        private set

    /**
     * @return Valid screens
     * @see NotificationData.setValidScreens
     */
    var validScreens: List<c<out AppComponent?>?>? = null
        private set

    /**
     * @return Block corresponding the onClick action of the icon in the top right corner of the notification.
     * @see NotificationData.setOnClickTopRightIcon
     */
    var onClickTopRightIcon: Function1<View?, Unit>? = null
        private set

    /**
     * @return Block corresponding the onClick action of the notification.
     * @see NotificationData.setTitle
     */
    var onClick: Function1<View?, Unit>? = null
        private set

    /**
     * Sets the title.
     * @param title Title.
     * @return [NotificationData] for chaining.
     */
    fun setTitle(title: CharSequence?): NotificationData {
        this.title = title
        return this
    }

    /**
     * Sets the subtitle.
     * @param subtitle Subtitle.
     * @return [NotificationData] for chaining.
     */
    fun setSubtitle(subtitle: CharSequence?): NotificationData {
        this.subtitle = subtitle
        return this
    }

    /**
     * Sets the body.
     * @param body Body.
     * @return [NotificationData] for chaining.
     */
    fun setBody(body: CharSequence?): NotificationData {
        this.body = body
        return this
    }

    /**
     * Sets the background for the attachment.
     * @param attachmentBackground Background [Drawable] of the attachment.
     * @return [NotificationData] for chaining.
     * @see NotificationData.setAttachment
     * @see NotificationData.setAttachmentUrl
     */
    fun setAttachmentBackground(attachmentBackground: Drawable?): NotificationData {
        this.attachmentBackground = attachmentBackground
        return this
    }

    /**
     * Sets the attachment URL.
     * @param attachmentUrl URL of the attachment.
     * @return [NotificationData] for chaining.
     */
    fun setAttachmentUrl(attachmentUrl: String?): NotificationData {
        this.attachmentUrl = attachmentUrl
        return this
    }

    /**
     * Sets the attachment.
     * @param attachment Attachment [Drawable].
     * @return [NotificationData] for chaining.
     */
    fun setAttachment(attachment: Drawable?): NotificationData {
        this.attachment = attachment
        return this
    }

    /**
     * Sets the stickers.
     * @param stickers [List] of stickers.
     * @return [NotificationData] for chaining.
     */
    fun setStickers(stickers: List<Sticker>?): NotificationData {
        this.stickers = stickers
        return this
    }

    /**
     * Sets the icon URL.
     * @param iconUrl URL of the icon.
     * @return [NotificationData] for chaining.
     */
    fun setIconUrl(iconUrl: String): NotificationData {
        this.iconUrl = iconUrl
        return this
    }

    /**
     * Sets the icon.
     * @param iconResId [IdRes] of the icon [Drawable].
     * @return [NotificationData] for chaining.
     */
    fun setIconResId(iconResId: Int?): NotificationData {
        this.iconResId = iconResId
        return this
    }

    /**
     * Sets the icon on the top right corner of the notification.
     * @param iconTopRight [Drawable] of the icon.
     * @return [NotificationData] for chaining.
     */
    fun setIconTopRight(iconTopRight: Drawable?): NotificationData {
        this.iconTopRight = iconTopRight
        return this
    }

    /**
     * Sets the auto dismiss period.
     * @param autoDismissPeriodSecs Dismiss period in seconds.
     * @return [NotificationData] for chaining.
     */
    fun setAutoDismissPeriodSecs(autoDismissPeriodSecs: Int?): NotificationData {
        this.autoDismissPeriodSecs = autoDismissPeriodSecs
        return this
    }

    /**
     * Sets valid screens for the notification.
     * @param validScreens Valid screens.
     * @return [NotificationData] for chaining.
     */
    fun setValidScreens(validScreens: List<c<out AppComponent?>?>?): NotificationData {
        this.validScreens = validScreens
        return this
    }

    /**
     * Sets the callback for the notification top right corner icon onClick action.
     * @param onClickTopRightIcon Block to execute after clicking the icon in the top right corner of the notification.
     * @return [NotificationData] for chaining.
     */
    fun setOnClickTopRightIcon(onClickTopRightIcon: Function1<View?, Unit>?): NotificationData {
        this.onClickTopRightIcon = onClickTopRightIcon
        return this
    }

    /**
     * Sets the callback for the notification onClick action.
     * @param onClick Block to execute after clicking the notification.
     * @return [NotificationData] for chaining.
     */
    fun setOnClick(onClick: Function1<View?, Unit>?): NotificationData {
        this.onClick = onClick
        return this
    }
}
